---
title: "Assignment 6: ATAC-seq"
output:
  github_document:
    toc: true
    toc_depth: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Assignment overview

*Today we will look at a differential ATAC-seq dataset between cells treated with an anti BAF protac and control (untreated) cells. The cell type is HAP1, a cancer cell line with a near-haploid genome. We will use this dataset to explore differential analysis. *

*The GEO entry is located here, where you can read more about the experiments: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE148175 *

*This is the paper: https://www.nature.com/articles/s41588-021-00777-3 *

*"Acute BAF perturbation causes immediate changes in chromatin accessibility"*



# Part 0: Getting ready 

```{r}
#install any of these you might not have already

library(ggplot2)
library(edgeR)
library(reshape)
library(GenomicRanges)
library(csaw)
library(Biostrings)


```


*ATAC-seq (Assay for Transposase-Accessible Chromatin using sequencing) produces reads which are generally in a fastq format, similar to the files you used in previous assignments. The reads in the fastq file are then aligned to the genome and a peak calling software is used to identify peaks. *

*In this assignment you will be downloading a counts matrix where each row corresponds to a peak (annotated by the genomic region) and each column corresponds to a treatment (control/no treatment, DMSO, BRM014 etc.) and timepoint and replicate. The value in a given cell corresponds to the number of reads for found for a chromatin region in at a given timepoint, treatment and replicate.*

*The file we will be using today can be found at the GEO entry. We have copied the relevant ftp link below: *

```{r}

#download the data
atacSeqData = read.table(textConnection(readLines(gzcon(url("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE148nnn/GSE148175/suppl/GSE148175_count_matrix_raw_atac_BRM014_ACBI1.csv.gz")))), 
                      sep=",", stringsAsFactors = FALSE, header = TRUE)

```


### `#?#` *Recall that each row is a peak and each column is a treatment/replicate/timepoint triplet. Report the value listed for chr1:25429953-25431146 in the BR_protac treatment after 6 hours. - 0.25 pt*
```{r}
# find the values corresponding to BR_protac after 6 hours for the peak at chr1:25429953-25431146 .

```

### `#?#` *Compare the value(s) with those for the control treatment at the same time and position. Are the counts higher or lower than the count for BR_protac? - 0.25 pt *
```{r}

```

### `#?#` *Does a high value for a given position in the count matrix indicate high accessibility or low accessibility for that region? - 0.5 pt* 
```{r}

```


*Now we will attempt to order the metadata contained within each of the various column headers of the counts matrix. We have provided the code for this below:*

```{r}

#create a sample metadata data.frame

samples = data.frame(ID = names(atacSeqData)[2:ncol(atacSeqData)])
samples$replicate = gsub("(R[12])_([0-9]+[minh]+)_(.*)$","\\1",samples$ID)
samples$timeName = gsub("(R[12])_([0-9]+[minh]+)_(.*)$","\\2",samples$ID)
samples$treatment = gsub("(R[12])_([0-9]+[minh]+)_(.*)$","\\3",samples$ID)
samples$treatment[samples$treatment=="N"]="BRM014"
samples$time= as.numeric(gsub("[a-z]*","",samples$timeName))
samples$time[grepl("min",samples$timeName)]=samples$time[grepl("min",samples$timeName)]/60

```


# Part 1: understanding the experiment

*Now using `samples` make a plot showing the experimental design, with time on the x axis, treatment on the y axis. If no sample for the time:treatment pair exists the point should be absent. Use the colour of the points to differentiate between presence in one replicate, the first replicate only or the second replicate only. *

### `#?#` *Make the above plot. Each point should represent one of the samples.  - 1 pt*
```{r}
#Remember: You should only have a point at a given time & position if such a sample exists in the dataframe. 

```

*In this study, one of the things they were comparing was BRM014 to DMSO. The drug BRM014 is dissolved in DMSO, so DMSO alone is the appropriate control to gauge the effect of BRM014.*

### `#?#` *Can we compare BRM014 to DMSO across all time points? Why/why not?  - 1 pt*

# Part 2: QC

*With most genomics data, it is important both that samples have sufficient coverage, and that the samples have similar coverage. Either case can lead to underpowered analysis, or misleading results. Calculate the read coverage for each sample. *

### `#?#` Make a plot with read coverage on the y-axis (total number of reads) and the samples on the x-axis. - 3 pt*

```{r}
# there are many ways you could do this; one of which is using the melt/cast functions from reshape

```

### `#?#` *Which sample has the most coverage? - 0.5 pt*
```{r}

```

### `#?#` *Which sample has the least? - 0.5 pt*
```{r}

```

### `#?#` *What is the % difference between the max and min (relative to the min)? - 0.5 pt*
```{r}

```

### `#?#` *Suppose you have two samples only one of which has sufficient coverage, but having significantly more coverage than the other. Briefly discuss the advantage(s) and disadvantage(s) of downsampling one to make the coverage more similar between samples. - 3 pt*


*For this assignment, we will look only at BI_protac vs control data. *

### `#?#` *Create a new data.frame containing only the BI_protac and control samples. - 1 pt*
```{r}


```

### `#?#` *For this subset, calculate the counts per million reads (CPM) for each sample - 2 pt*
```{r}

```


### `#?#` *Plot the kernel density estimate for CPM (x axis). 1 curve per sample, different colours per curve. - 1 pt*

```{r}

```

### `#?#` *Plot the kernel density estimate for log(CPM+1) (x axis), coloured as before - 1 pt*

```{r}

```

### `#?#` *Why do you think log-transforming is usually performed when looking at genomics data? - 1 pt*

### `#?#` *Why do you think we add 1 before log transforming? - 1 pt* 

### `#?#` *Some regions have very large CPMs. Seperate the peaks for which CPM>300. -1.5 pt*
```{r}

```

### `#?#` *Make a bar plot of the number peaks for which CPM>300 found on each chromosome. What do you notice? - 1.5 pt*
```{r}
```

*Normally, we would remove some of these regions before continuing (and would redo the above steps). Since this is an assignment, we will continue with the data as-is.*

*Often a good first step is to see if the data look good. One way to do this is by seeing whether or not the signals in each sample correlate with each other in ways you expect.*

### `#?#` *Calculate the pairwise correlations between log(CPM+1)s for the samples and plot them as a heatmap (samples x samples) - 3 pt*
```{r}


```

### `#?#` *What do you expect the correlations between replicates to look like? Is that what you see? - 2 pt*



*It is common to exclude some regions from analysis. For instance, we won't be able to robustly identify those that are differential but have low coverage even if they are truly differential, so there is no point testing these. We will also remove mitochondrial regions, a common contaminant of ATAC-seq data.*

### `#?#` *Filter your data, retaining only regions where the average counts per sample is greater than 20, and also remove mitochondrial regions - 3 pt*
```{r}




```

### `#?#` *How many peaks did you have before? How many do you have now? - 1 pt*
```{r}

```

# Part 3: Differential ATAC

*We want to know what regions are differentially accessible between BI_protac and the control.* 

*Today, we're going to use edgeR, which is designed for RNA-seq, but works well on ATAC-seq as well. The user guide is here:* https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf


### `#?#` *Make a count matrix called `countMatrix` for the BI_protac and control samples, including only the peaks we retained above - 2 pt*
```{r}

```

*EdgeR is exceptionally versatile, with many different options for analysis. Today, you're going to use the GLM-quasi-likelihood approach to calculate differential accessibility. We are providing the first example analysis below, which you can modify in subsequent steps. You will need to understand what the steps do, so read the appropriate documentation. *
```{r}
curSamples = samples[match(names(countMatrix), samples$ID),];
y = DGEList(counts=countMatrix, group=curSamples$treatment)
y = calcNormFactors(y)
designPaired = model.matrix(~curSamples$treatment + curSamples$timeName)  
# we are using timeName here to make sure that time is treated as a categorical variable. Had we more time points it might make sense to treat time as a value.
y = estimateDisp(y, designPaired)
fitPaired = glmQLFit(y, designPaired)
qlfPairedTime6vs24 = glmQLFTest(fitPaired, coef=3) 
qlfPairedTreatControlvsProtac = glmQLFTest(fitPaired, coef=2)

allDEStatsPairedTreatControlvsProtac = as.data.frame(topTags(qlfPairedTreatControlvsProtac,n=nrow(countMatrix)))
allDEStatsPairedTreatControlvsProtac$region=row.names(allDEStatsPairedTreatControlvsProtac)

allDEStatsPairedTime6vs24 = as.data.frame(topTags(qlfPairedTime6vs24,n=nrow(countMatrix)))
allDEStatsPairedTime6vs24$region=row.names(allDEStatsPairedTime6vs24)

```
*While the differential analysis has been done in this case, before we look at the results, we are going to check if the data appear to be normalized correctly. Also include a loess line of best fit, and the line y=0.*

### `#?#` *Make an MA plot for allDEStatsPairedTreatControlvsProtac -2pt*
```{r}

```

### `#?#` *Make an MA plot for allDEStatsPairedTime6vs24 - 1 pt*
```{r}

```

### `#?#` *Briefly explain what you are looking at. What do the axes show? What is the loess line for? - 2 pts*

*Now we're going to test loess normalization instead.* 


### `#?#` *Perform the same differential peak analysis using loess regularization. - 1 pt*
```{r}
#Note: the Bioconductor package csaw implements loess regularization in a way that is compatible with edgeR
## Tip: use the csaw library to implement the loess regularization


```

### `#?#` *Make the same two MA plots as before, but this time using the loess normalized analysis - 1 pt*
```{r}



```

### `#?#` *What was the first normalization method? What changed in the MA plots? Which analysis do you think is more reliable and why? - 4 pt*
*Hint: Think about what loess normalization corrects for.*


# Part 4: GC bias

*Next, we will look at potential GC bias in the data. We will again use bioconductor *

### `#?#` *Convert the region IDs to a GRanges object - 3 pt*
```{r}
#note that the names of your peaks are of the format <chr>:<startPos>-<endPos>
## Tip: lookinto the GenomicRanges documentation 



```


### `#?#` *Extract the genomic DNA sequences for each peak using hg38 - 3 pt*
*See for relevant documentation: https://bioconductor.org/packages/release/workflows/vignettes/sequencing/inst/doc/sequencing.html *
```{r}
## Tip: Use the Biostring library 
library(BSgenome.Hsapiens.UCSC.hg38)

```


*Now we will see if there's any relationship between peak CPM and GC content for each of the samples.*

### `#?#` *Create scatter plots (one per sample, e.g. using facet_wrap), including lines of best fit (GAM), where each plot shows GC content (x axis) vs CPM (y axis) for each peak (points) -2pt*
```{r}
#please limit the y axis to between 0 and 50

```

### `#?#` *Repeat the above, but this time showing only the lines of best fit and all on the same plot - 2 pt*
```{r}

```


### `#?#` *Given this result, predict whether we will see a significant relationship between GC content and logFC in our differential peak analysis (loess-normalized). Justify your prediction. Predicting "wrong" will not be penalized, as long as your justification is correct. Don't retroactively change your answer. - 2 pt*





### `#?#` *Plot the relationship between GC and logFC for the loess-normalized ControlvsProtac analysis. Also include a line of best fit (blue) and y=0 (red) - 2 pt*
```{r}

```

### `#?#` *Now plot the same thing for the NON loess-normalized ControlvsProtac analysis. - 1 pt*
```{r}


```


### `#?#` *Was your prediction correct? Do you think we should also account for GC normalization in our differential ATAC analysis? Why/why not? - 3 pt*



*We will leave GC normalization as an optional exercise, and will not actually do it here.*

# Part 5: Differential analysis results

### `#?#` *Suppose we perform the analyses above, redoing the differential analysis once more with GC normalization, and also considering that we tested loess and the default normalization methods. Did we P-hack? Why or why not? - 2 pt*



*Going forward, we will only use the initial analysis (**not loess normalized**)*

### `#?#` *Now considering the two comparisons (6 vs 24 hours, and protac vs control). EdgeR performed a correction for MHT, but if we want to analyze the results from both comparisons, do we need to re-adjust to account for the fact that we tested two different hypothesis sets (time and treatment)? Why/not? - 2 pt*




### `#?#` *How many differential peaks did you find (FDR<0.01). - 1 pt*
```{r}


```

### `#?#` *Make a volcano plot of the allDEStatsPairedTreatControlvsProtac, with -log10(p-value) on the y axis and logFC on the x. Colour points that are significant at an FDR<0.01. - 2 pt*
```{r}


```




### `#?#` *Plot the logCPM (x axis) by -log10(Pvalue) (y axis), again colouring by FDR<0.01. - 2 pt*
```{r}


```

### `#?#` *Do you think our initial filtering on peaks with at least 10 reads on average per sample was a good choice? Why or why not?*



*At this point there are many other follow ups you can and would do for a real differential analysis, but we leave these as optional exercises. For example:*
1. Confirming that the differential peaks look correct (e.g. CPM heatmap)
2. Confirming that peaks look differential on the genome browser
3. Looking for motif enrichment
4. Performing a GREAT analysis, including functional enrichment and assigning peaks to genes

*Knit your assignment as a github_document and submit the resulting .md and this .Rmd to your github, and complete the assignment submission on Canvas. Make sure to include the graphs with your submission. *
 



# Authors and contributions

Following completion of your assignment, please fill out this section with the authors and their contributions to the assignment.  If you worked alone, only the author (e.g. your name and student ID) should be included.

Authors: Name1 (studentID1) and Name2 (studentID2)

Contributions: (example) N1 and N2 worked together on the same computer to complete the assignment. N1 typed for the first half and N2 typed for the second half. 